File: wpimath/src/main/native/cpp/geometry/struct/Twist2dStruct.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "wpi/math/geometry/struct/Twist2dStruct.hpp"

namespace {
constexpr size_t DX_OFF = 0;
constexpr size_t DY_OFF = DX_OFF + 8;
constexpr size_t DTHETA_OFF = DY_OFF + 8;
}  // namespace

using StructType = wpi::util::Struct<wpi::math::Twist2d>;

wpi::math::Twist2d StructType::Unpack(std::span<const uint8_t> data) {
  return wpi::math::Twist2d{
      wpi::units::meters<>{wpi::util::UnpackStruct<double, DX_OFF>(data)},
      wpi::units::meters<>{wpi::util::UnpackStruct<double, DY_OFF>(data)},
      wpi::units::radians<>{wpi::util::UnpackStruct<double, DTHETA_OFF>(data)},
  };
}

void StructType::Pack(std::span<uint8_t> data,
                      const wpi::math::Twist2d& value) {
  wpi::util::PackStruct<DX_OFF>(data, value.dx.value());
  wpi::util::PackStruct<DY_OFF>(data, value.dy.value());
  wpi::util::PackStruct<DTHETA_OFF>(data, value.dtheta.value());
}

File: wpimath/src/main/java/org/wpilib/math/geometry/struct/Twist2dStruct.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package edu.wpi.first.math.geometry.struct;

import edu.wpi.first.math.geometry.Twist2d;
import edu.wpi.first.util.struct.Struct;
import java.nio.ByteBuffer;

public class Twist2dStruct implements Struct<Twist2d> {
  @Override
  public Class<Twist2d> getTypeClass() {
    return Twist2d.class;
  }

  @Override
  public String getTypeName() {
    return "Twist2d";
  }

  @Override
  public int getSize() {
    return kSizeDouble * 3;
  }

  @Override
  public String getSchema() {
    return "double dx;double dy;double dtheta";
  }

  @Override
  public Twist2d unpack(ByteBuffer bb) {
    double dx = bb.getDouble();
    double dy = bb.getDouble();
    double dtheta = bb.getDouble();
    return new Twist2d(dx, dy, dtheta);
  }

  @Override
  public void pack(ByteBuffer bb, Twist2d value) {
    bb.putDouble(value.dx);
    bb.putDouble(value.dy);
    bb.putDouble(value.dtheta);
  }
}
